pkg/flavor/flavor.go

```diff
@@ -44,6 +52,213 @@ func NewMethods(clientset kubecli.KubevirtClient) Methods {
 	}
 }
 
+func GetRevisionName(vmName, resourceName string, resourceUID types.UID, resourceGeneration int64) string {
+	return fmt.Sprintf("%s-%s-%s-%d", vmName, resourceName, resourceUID, resourceGeneration)
+}
+
+func CreateFlavorControllerRevision(vm *virtv1.VirtualMachine, revisionName string, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec) (*appsv1.ControllerRevision, error) {
+
+	revisionPatch, err := json.Marshal(flavorSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            revisionName,
+			Namespace:       vm.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(vm, virtv1.VirtualMachineGroupVersionKind)},
+		},
+		Data: runtime.RawExtension{Raw: revisionPatch},
+	}, nil
+
+}
+
+func (m *methods) createFlavorRevision(vm *virtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+
+	switch strings.ToLower(vm.Spec.Flavor.Kind) {
+	case apiflavor.SingularResourceName, apiflavor.PluralResourceName:
+		flavor, err := m.findFlavor(vm)
+		if err != nil {
+			return nil, err
+		}
+		revision, err := CreateFlavorControllerRevision(vm, GetRevisionName(vm.Name, flavor.Name, flavor.UID, flavor.Generation), &flavor.Spec)
+		if err != nil {
+			return nil, err
+		}
+		return revision, nil
+
+	case apiflavor.ClusterSingularResourceName, apiflavor.ClusterPluralResourceName:
+		clusterFlavor, err := m.findClusterFlavor(vm)
+		if err != nil {
+			return nil, err
+		}
+
+		revision, err := CreateFlavorControllerRevision(vm, GetRevisionName(vm.Name, clusterFlavor.Name, clusterFlavor.UID, clusterFlavor.Generation), &clusterFlavor.Spec)
+		if err != nil {
+			return nil, err
+		}
+		return revision, nil
+	default:
+		return nil, fmt.Errorf("got unexpected kind in FlavorMatcher: %s", vm.Spec.Flavor.Kind)
+	}
+}
+
+func (m *methods) storeFlavorRevision(vm *virtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+
+	if vm.Spec.Flavor == nil || len(vm.Spec.Flavor.RevisionName) > 0 {
+		return nil, nil
+	}
+
+	flavorRevision, err := m.createFlavorRevision(vm)
+	if err != nil {
+		return nil, err
+	}
+
+	flavorRevision, err = m.clientset.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), flavorRevision, metav1.CreateOptions{})
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
```

Comment:
The existing ControllerRevision may have different `Data` from what we expect. Can you check the `Data` field?

Reply:
I'm going to handle this in a follow up commit *after* the other code fetching controller revisions so I can reuse code from there.

Reply:
Okay this is done in fa9c46c8da6c3d2f07ada10a70f1ca26e610da9e now.